### PR TITLE
Migrate Improve > Shipping > Preferences controller

### PR DESCRIPTION
**Pull request by rokaszygmantas**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migration of Shipping > Preferences controller (legacy AdminShipping)
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | Compare to the legacy controller.

> Needs https://github.com/PrestaShop/PrestaShop/pull/9120 to be merged first 

> Note: I'm concerned about the code in AdminShippingController::postProcess(). It looks like the code in that method is never used - the if condition is never satisfied ```if (Tools::isSubmit('submitFees'.$this->table)) {``` the post key is 'submitFeesdelivery' and there is no such key in html or in other files.
The code I'm referring to:
https://github.com/PrestaShop/PrestaShop/blob/70b7e25a902ebdcb91e8a9463d12480a3f2827c9/controllers/admin/AdminShippingController.php#L126
Would be great to get some input on that.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9094)

### CODE REVIEW

File: src/Adapter/Carrier/CarrierDataProvider.php [hunk 1]
```diff
@@ -27,6 +27,7 @@
 namespace PrestaShop\PrestaShop\Adapter\Carrier;
 
 use Carrier;
+use Configuration;
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/CarrierDataProvider.php]: you need to use Configuration from Adapter namespace

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/CarrierOptionsType.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        $this->orderWayChoices = $orderWayChoices;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $carrierChoices = array_merge([
+            'Best price' => -1,
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/CarrierOptionsType.php:76]: will these labels be translated?

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/PreferencesFormDataProvider.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    public function setData(array $data)
+    {
+        /**
+         * All three following fields (shipping_handling_charges, free_shipping_price, free_shipping_weight)
+         * can be submitted as either empty or with "0" values
+         * and in both cases it means that the functionality is disabled.
+         * For consistency changing the values to 0, in case they were submitted empty.
+         */
+        $data['handling']['shipping_handling_charges'] = $data['handling']['shipping_handling_charges'] ?: 0;
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/PreferencesFormDataProvider.php]: What you want to do here, is to declare a default value if empty: you can do it in your form ;)

```php
$builder
    ->add('shipping_handling_charges', MoneyWithSuffixType::class, [
            'currency' => $defaultCurrency->iso_code,
            'suffix' => $this->trans('(tax excl.)', 'Admin.Global'),
            'required' => false,
            'empty_data' => 0,
        ]);
```

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+{% extends '@PrestaShop/Admin/layout.html.twig' %}
+
+{% set handlingForm, carrierOptionsForm = form.handling, form.carrier_options %}
+
+{% block content %}
+    <div class="container">
+        {{ form_start(form, {'attr': {'class': 'form', 'id': 'configuration_form'}}) }}
+        <div class="row">
+            {# Handling options block #}
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig]: this comment is wrong, isn't?

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]
minor fixes and improvements to be done, good job Rokas!

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/CarrierOptionsType.php (hunk 2 above)
Comment (rokaszygmantas) [reply src/PrestaShopBundle/Form/Admin/Improve/Shipping/Preferences/CarrierOptionsType.php:76]: yes :) https://prnt.sc/jlztqm

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig (hunk 4 above)
Comment (rokaszygmantas) [reply src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig]: I don't think it's wrong - the options block is called "Handling", but it can sound differently so I'll rephrase it to avoid double meanings

File: src/Adapter/Carrier/CarrierDataProvider.php [hunk 5]
```diff
@@ -55,6 +56,55 @@ public function getCarriers($id_lang, $active = false, $delete = false, $id_zone
...
     }
 
+    /**
+     * Get all active carriers in given language, usable for choice form type
+     *
+     * @param int|null $languageId if not provided - will use the default language
+     *
+     * @return array carrier choices
+     */
+    public function getActiveCarriersChoices($languageId = null)
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/CarrierDataProvider.php:76]: How about inject configuration from adapter namespace in constructor of this class and rely on it instead?

Could be easier to unit test :)

File: src/Adapter/Carrier/CarrierOptionsConfiguration.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Carrier;
+
+use PrestaShop\PrestaShop\Adapter\Configuration;
+use PrestaShop\PrestaShop\Core\Configuration\DataConfigurationInterface;
+
+class CarrierOptionsConfiguration implements DataConfigurationInterface
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/CarrierOptionsConfiguration.php]: miss PHPDoc

File: src/Adapter/Carrier/CarrierOptionsConfiguration.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+{
+    /**
+     * @var Configuration
+     */
+    private $configuration;
+
+    public function __construct(Configuration $configuration)
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/CarrierOptionsConfiguration.php]: miss PHPDoc (you can use the generator of PHPStorm)

File: src/Adapter/Carrier/HandlingConfiguration.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Carrier;
+
+use PrestaShop\PrestaShop\Adapter\Configuration;
+use PrestaShop\PrestaShop\Core\Configuration\DataConfigurationInterface;
+
+class HandlingConfiguration implements DataConfigurationInterface
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/HandlingConfiguration.php]: miss PHPDoc

File: src/Adapter/Carrier/HandlingConfiguration.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+{
+    /**
+     * @var Configuration
+     */
+    private $configuration;
+
+    public function __construct(Configuration $configuration)
```
Comment (mickaelandrieu) [inline src/Adapter/Carrier/HandlingConfiguration.php]: miss PHPDoc

File: src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig [hunk 10]
```diff
@@ -31,10 +31,10 @@
     <div class="container">
         {{ form_start(form, {'attr': {'class': 'form', 'id': 'configuration_form'}}) }}
         <div class="row">
-            {# Handling options block #}
+            {# Block for "Handling" fieldset #}
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/Improve/Shipping/Preferences/preferences.html.twig]: thanks :)

**Comment by rokaszygmantas** [comment]
@mickaelandrieu thanks for the comments. I've addressed them all now :)

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by prasanthSelvar** [comment]
Hello guys,

To change some parameter in Shipping preferences, I must first change the value _free shipping starts at_ to another value than 0 kg. If not I have an invalid error.
In fact, I cannot change the value of _free shipping starts at_ to 0 kg after a modification.

This thing works fine on 1.7.3.3.

**Comment by mickaelandrieu** [comment]
ping @kpodemski, are you aware of something about the code in ``AdminShippingController::postProcess()``, I can't see any use case where this code is used. If it's dead code it's time to remove it, but if it's not I'd like a short explanation on what it does, I can't find any information about it :/

**Comment by kpodemski** [comment]
@mickaelandrieu @prasanthSelvar 

unless this code somehow is not used by Carrier Wizard feel free to remove it, it has been in this file for years so i guess it is some old code frome times before refactor of setting up prices and zones for Carriers

**Comment by rokaszygmantas** [comment]
I'll get back to this PR since https://github.com/PrestaShop/PrestaShop/pull/9120 is already merged

**Comment by mickaelandrieu** [comment]
Thanks @rokaszygmantas :)

**Comment by rokaszygmantas** [comment]
@mickaelandrieu rebased - looks like the _free shipping starts at_ field is working as expected now :)

**Comment by mickaelandrieu** [comment]
@rokaszygmantas yeah I need to find someone in the QA team to validate

**Review by mickaelandrieu — APPROVED** [review]

**Review by mickaelandrieu — APPROVED** [review]

**Comment by marionf** [comment]
@rokaszygmantas & @mickaelandrieu 
Everything is good :+1:

**Comment by PierreRambaud** [comment]
Thanks @rokaszygmantas & @mickaelandrieu and of course @marionf for the QA feedback \o/

**Comment by robertomat** [comment]
Can you help me about this issue??? i´m going crazy
Template "@PrestaShop/Admin/Improve/Shipping/Preferences/preferences.html.twig" is not defined.


CRITICAL13:09:40 | php | Uncaught Exception: Template "@PrestaShop/Admin/Improve/Shipping/Preferences/preferences.html.twig" is not defined.
-- | -- | --
CRITICAL13:09:40 | request | Uncaught PHP Exception InvalidArgumentException: "Template "@PrestaShop/Admin/Improve/Shipping/Preferences/preferences.html.twig" is not defined." at /var/www/vhosts/awasunglasses.com/prestawa/vendor/symfony/symfony/src/Symfony/Bridge/Twig/TwigEngine.php line 127
